### PR TITLE
Alternative design for exception translators

### PR DESCRIPTION
**Pull request by pschella**
This design makes the default handing into just another exception translator which is always tried last. With this design user registered translators are no longer required to always catch all exceptions (which they may forget to do, breaking the exception translation chain). Instead they should only catch those exceptions they actually translate, leaving the rest to the next registered translator.

A downside of this design is that exception handling is perhaps less readable, because it is split up into different files and objects.

### CODE REVIEW

File: include/pybind11/exception_translator.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+NAMESPACE_BEGIN(pybind11)
+
+template <typename ExceptionTranslator>
+void register_exception_translator(ExceptionTranslator translator) {
+    detail::get_internals().registered_exception_translators.push_front(std::move(translator));
+}
+
+//NAMESPACE_BEGIN(detail)
+//NAMESPACE_END(detail)
```
Comment (TallJimbo) [inline include/pybind11/exception_translator.h]: Looks like these should be removed?

File: include/pybind11/pybind11.h [hunk 2]
```diff
@@ -338,6 +338,7 @@ class cpp_function : public function {
 
         handle parent = nargs > 0 ? PyTuple_GET_ITEM(args, 0) : nullptr,
                result = PYBIND11_TRY_NEXT_OVERLOAD;
+
```
Comment (TallJimbo) [inline include/pybind11/pybind11.h]: Looks like an unnecessary change.

File: include/pybind11/pybind11.h [hunk 3]
```diff
@@ -394,19 +395,19 @@ class cpp_function : public function {
...
+            return nullptr;
         } catch (...) {
-            PyErr_SetString(PyExc_RuntimeError, "Caught an unknown exception!");
+            auto &registered_exception_translators = pybind11::detail::get_internals().registered_exception_translators;
+            for (auto& translator : registered_exception_translators) {
+                try {
+                    translator(std::current_exception());
+                } catch (...) {
+                    continue;
+                }
```
Comment (TallJimbo) [inline include/pybind11/pybind11.h:419]: I think it might be better to have translators return a `bool` rather than throw to indicate that they're not handling the exception, in keeping with the principle that we should only use exceptions for unusual conditions.  But it'd probably still be good to wrap this in another try block so translators can throw exceptions if they encounter genuine errors.

File: include/pybind11/pybind11.h [hunk 4]
```diff
@@ -394,19 +395,19 @@ class cpp_function : public function {
...
+                return nullptr;
+            }
+            PyErr_SetString(PyExc_RuntimeError, "Cannot translate exception!");
```
Comment (TallJimbo) [inline include/pybind11/pybind11.h]: I'd recommend `PyExc_SystemError` instead of `RuntimeError`, I think, given that if we've gotten to this point Pybind11's standard translators have fired.

File: include/pybind11/exception_translator.h (hunk 1 above)
Comment (pschella) [inline include/pybind11/exception_translator.h]: Why? Users still need a way to register exception translators. `get_interrnals` is not exposed and clumsy to use. The only reason for not using this function to do the default registration is to avoid the need for including this header.

File: include/pybind11/pybind11.h (hunk 3 above)
Comment (pschella) [inline include/pybind11/pybind11.h:419]: I'm not sure. The current behaviour is for the exception to "fall through" to the next exception translator if it is not handled. This is equivalent to the situation where all exceptions were caught in one big series of catch blocks. Thus an exception handler indicates it cannot handle something simply by doing nothing. Moreover, reintroducing a boolean return value would create quite a bit of control flow overhead on both the user and the pybind11 side (each has to catch all exceptions and set/check the return value).

File: include/pybind11/pybind11.h (hunk 4 above)
Comment (pschella) [inline include/pybind11/pybind11.h]: Agreed

File: include/pybind11/pybind11.h (hunk 2 above)
Comment (pschella) [inline include/pybind11/pybind11.h]: yep

File: include/pybind11/exception_translator.h (hunk 1 above)
Comment (pschella) [inline include/pybind11/exception_translator.h]: Ah, you meant the comments :) Yes, they should go.

File: include/pybind11/pybind11.h (hunk 3 above)
Comment (TallJimbo) [inline include/pybind11/pybind11.h:419]: Oh, I see.  You need to call `std::rethrow_exception` within the translators anyway so you can use a `try/catch` block to find out what it is.  I agree that makes my proposal unworkable, at least at this level.

But I am still a bit worried about the performance implications for the likely scenario in LSST, where every package that adds an exception would have to add a custom handler to the list, and that results in a lot of re-throwing of exceptions.  I suppose we could have a single custom handler for LSST, though, and have another way to register lightweight handlers there (since all of our exceptions inherit from a common base class and have virtual destructors, we could traverse the inheritance graph with dynamic casts, for instance).

As a side note: I'm not 100% sure what `std::current_exception` will point to here if the translator throws a different exception than the one it received (is it the last exception thrown, or the one in the currently active `catch` block`).  I suspect it must be the latter, but it might be good to clarify that by assigning the result of`std::current_exception` to a variable with the scope you want.
